sdk/core/azure-core/src/main/java/com/azure/core/models/CloudEvent.java

```diff
@@ -266,36 +250,44 @@ public final class CloudEvent {
-     *               for each distinct event.
+     * for each distinct event.
      * @param type Type of event related to the originating occurrence.
      * @param data A {@link BinaryData} that wraps the original data, which can be a String, byte[], or model class.
      * @param format Set to {@link CloudEventDataFormat#BYTES} to serialize the data to base64 format, or
-     *               {@link CloudEventDataFormat#JSON} to serialize the data to JSON value.
-     * @param dataContentType The content type of the data. It has no impact on how the data is serialized but tells
-     *                        the event subscriber how to use the data. Typically the value is of MIME types such as
-     *                        "application/json", "text/plain", "text/xml", "avro/binary", etc. It can be null.
-     * @throws NullPointerException if source, type is null, or format is null while data isn't null.
+     * {@link CloudEventDataFormat#JSON} to serialize the data to JSON value.
+     * @param dataContentType The content type of the data. It has no impact on how the data is serialized but tells the
+     * event subscriber how to use the data. Typically, the value is of MIME types such as "application/json",
+     * "text/plain", "text/xml", "avro/binary", etc. It can be null.
+     * @throws NullPointerException If source or type is null or format is null while data isn't null.
      * @throws IllegalArgumentException if format is {@link CloudEventDataFormat#JSON} but the data isn't in a correct
      * JSON format.
      */
-    public CloudEvent(String source, String type, BinaryData data, CloudEventDataFormat format, String dataContentType) {
-        Objects.requireNonNull(source, "'source' cannot be null.");
-        Objects.requireNonNull(type, "'type' cannot be null.");
-        this.source = source;
-        this.type = type;
+    public CloudEvent(String source, String type, BinaryData data, CloudEventDataFormat format,
+        String dataContentType) {
+        this.source = Objects.requireNonNull(source, "'source' cannot be null.");
+        this.type = Objects.requireNonNull(type, "'type' cannot be null.");
+
         if (data != null) {
             Objects.requireNonNull(format, "'format' cannot be null when 'data' isn't null.");
             if (CloudEventDataFormat.BYTES == format) {
                 this.dataBase64 = Base64.getEncoder().encodeToString(data.toBytes());
             } else {
-                try {
-                    this.data = BINARY_DATA_OBJECT_MAPPER.readTree(data.toBytes());
+                try (JsonReader jsonReader = JsonProviders.createReader(data.toBytes())) {
+                    JsonToken token = jsonReader.nextToken();
+                    if (token == JsonToken.START_OBJECT || token == JsonToken.START_ARRAY) {
+                        this.data = jsonReader.readChildren();
```

Comment:
Good catch, this runs into issue as `azure-json` doesn't do reflective serialization and it would mean complex types, not supported by JsonWriter, would serialize using their `toString` value. Will need to think this over on how to handle this.

JsonReader is less of an issue as this will read the JSON as the specification defines.

For what was mentioned here this is reading data as raw JSON whereas before it was reading it as a Jackson JsonNode/TreeNode. I think this implementation is better as we were using our deserializer for this which could be opinionated and change how the raw JSON was being read. This leaves it as raw JSON for BinaryData to interpret later with whatever `JsonSerializer` implementation the customer provides, or the default implementation which is our opinions.